src/main/java/io/pravega/perf/PerfStats.java

```diff
@@ -10,52 +10,61 @@
 package io.pravega.perf;
 
 import java.io.IOException;
-import java.util.Arrays;
-import java.util.concurrent.CompletableFuture;
-import java.time.Instant;
-import java.time.Duration;
-import java.util.concurrent.atomic.AtomicLong;
+import java.util.ArrayList;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ForkJoinPool;
+import java.util.concurrent.locks.LockSupport;
 import java.nio.file.Files;
 import java.nio.file.Paths;
 
 import org.apache.commons.csv.CSVFormat;
 import org.apache.commons.csv.CSVPrinter;
+import org.apache.commons.csv.CSVParser;
+import org.apache.commons.csv.CSVRecord;
 
 
 /**
  * class for Performance statistics.
  */
 public class PerfStats {
-    private static final long NANOSEC_PER_SEC = 1000000000L;
-    final private int messageSize;
     final private String action;
-    final private Instant start;
-    private long[] latencies;
-    private int sampling;
-    private int index;
-    private long count;
-    private long bytes;
-    private long maxLatency;
-    private long totalLatency;
+    final private String csvFile;
+    final private int messageSize;
     final private long windowInterval;
-    private timeWindow window;
-    final private AtomicLong eventID;
-    final private CSVPrinter printer;
+    private TimeWindow window;
```

Comment:
This is being accessed in multiple threads and is non-final. See comments below highlighting the race condition.

Reply:
now, this  timewindow and latency recorder are as part of QueueProcessor object.